src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/R2RReader.cs

```diff
@@ -298,16 +212,99 @@ public R2RReader(IAssemblyResolver assemblyResolver, MetadataReader metadata, PE
         /// </summary>
         /// <param name="filename">PE image</param>
         /// <exception cref="BadImageFormatException">The Cor header flag must be ILLibrary</exception>
-        public R2RReader(IAssemblyResolver assemblyResolver, string filename)
-            : base(assemblyResolver, filename, new List<string>())
+        public unsafe R2RReader(IAssemblyResolver assemblyResolver, string filename)
         {
+            _assemblyResolver = assemblyResolver;
+            Filename = filename;
             Initialize();
         }
 
         private unsafe void Initialize()
         {
-            IsR2R = ((PEReader.PEHeaders.CorHeader.Flags & CorFlags.ILLibrary) != 0);
-            if (!IsR2R)
+            _assemblyCache = new Dictionary<int, MetadataReader>();
+            this._manifestReferences = new List<AssemblyReferenceHandle>();
+
+            if (MetadataReader == null)
+            {
+                Image = File.ReadAllBytes(Filename);
+
+                fixed (byte* p = Image)
+                {
+                    IntPtr ptr = (IntPtr)p;
+                    PEReader = new PEReader(p, Image.Length);
```

Comment:
Same comment than below, taking a pointer on something that will be collected will likely lead to a memory corruption later when accessing MetadataReader

Reply:
The difference is that Image seems to be a field member, so the corruption will come after the byte array has been relocated by the GC

Reply:
Tracked in https://github.com/dotnet/runtime/issues/1219